sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala

```diff
@@ -844,6 +844,160 @@ abstract class OrcSourceSuite extends OrcSuite with SharedSparkSession {
       }
     }
   }
+
+  // Test the various cases during deserialization where we may or may not copy
+  // a struct writer's result row
+  test("struct in an array") {
+    withTempPath { file =>
+      val df = sql(
+        """SELECT
+          |  array(
+          |    named_struct(
+          |      'a1', 1,
+          |      'a2', 2),
+          |    named_struct(
+          |      'a1', 3,
+          |      'a2', 4)
+          |  ) as col1
+          |""".stripMargin)
+      df.write.orc(file.getCanonicalPath)
+      val df2 = spark.read.orc(file.getCanonicalPath)
+      checkAnswer(df2, df.collect().toSeq)
```

Comment:
Could you test both vectorized and non-vectorized code path? i.e. enabling and disabling `spark.sql.orc.enableNestedColumnVectorizedReader`? The current test case is working, because vectorized read of struct is disabled, so we are going through the non-vectorized code path (the deserialization code path we are fixing here). I am worried about missing this test, when we decide to enable vectorized reader for nested column in upcoming future Spark release. Thanks.